### DISCUSSION TITLE
Symbol search for C++?

### DISCUSSION THREAD

**Discussion by jedieaston** [community/community#8594] [Other Feature Feedback, Questions, & Ideas]
C++ isn't currently supported, but there is a tree-sitter plugin for it (https://github.com/tree-sitter/tree-sitter-cpp). Visual Studio's built in search is slow, but this new Code Search seems to work really well so far, especially when symbols are supported. 

Is there any way this can make it on the roadmap?

**Comment by dcreager** [comment]
It's definitely on our radar to support symbol extraction from more languages.  C++ is a bit more difficult than other languages because of the preprocessor and template language, but still should be doable.  Our symbol extraction builds on tree-sitter's [`tags`](https://github.com/tree-sitter/tree-sitter/tree/master/tags) feature, so if we can get the `tree-sitter tags` command to spit out C++ symbols locally, then we can fold it into our code search indexes and code navigation.

**Comment by tclem** [comment]
We have actually tested with that c++ parser, but disabled it right before releasing the tech preview because it is the most resource intensive parser and has a very high failure and timeout rate. Our goal is to fully support this language, but we need to spend a little bit more time making that parser work at our scale.

**Reply by jdrouhard** [reply]
Which version did you test? Do you have repros for source files that have high failure rates and/or timeouts?

**Reply by colinwm** [reply]
We were on 0.20.0. I'm not sure if we really looked into C vs. C++ parser performance, but IIRC the linux kernel has a lot of pretty hard-to-parse C files.

**Reply by jdrouhard** [reply]
Heavy macro-laden c/c++ won't really be parseable by treesitter, unfortunately. But if you're seeing slow performance on "normal" code (without a lot of macros), send me some examples. I'd like to see these languages get supported!

**Reply by colinwm** [reply]
Yeah, we absolutely want to support it... when I get some time to look into this more deeply, I'll provide some feedback in an issue on your repo and maybe we can figure it out.

**Reply by gevorgyana** [reply]
Do you hire interns to do that?

**Reply by colinwm** [reply]
@gevorgyana - we're not hiring interns right now AFAICT, but if you want to help us solve this problem (and contribute to open source!), we're using https://github.com/tree-sitter/tree-sitter-cpp to parse C/C++. The library is very slow, though, perhaps because C++ is extremely complex to parse.

The best route, I think, would be to try to improve performance - either within that library, or by forking it and creating a stripped-down parser that supports, e.g. 80% of C++ features, but is ~10x faster. If somebody were to implement that, we would definitely use that parser to index C/C++. If nobody implements it, we'll probably do it "eventually"...
